crates/web-sys/tests/wasm/element.js

```diff
@@ -151,6 +151,14 @@ export function new_title() {
   return document.createElement("title");
 }
 
+export function new_webgl_rendering_context() {
+  const foo = document.createElement('canvas');
+  console.log('Does get context work? ' + foo.getContext('webgl'));
+
+  const canvas = document.createElement('canvas');
+  return canvas.getContext('webgl');
```

Comment:
@alexcrichton alright hit another roadblock.

It looks like I'm not actually creating a `WebGlRenderingContext` but am instead getting `null` back when I call `getContext('webgl')` on a canvas.

```
failures:

---- wasm::whitelisted_immutable_slices::test_webgl_rendering_context_immutable_slices output ----
    log output:
        Does get context work? null

    error output:
        wasm-bindgen: imported JS function that was not marked as `catch` threw an error: TypeError: getObject(...) is null

    JS exception that was thrown:
        TypeError: getObject(...) is null
        __widl_f_vertex_attrib1fv_with_f32_array_WebGLRenderingContext@http://127.0.0.1:56658/wasm-bindgen-test:5119:9
        web_sys::WebGlRenderingContext::vertex_attrib1fv_with_f32_array::h52eb674a885910ee@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[3417]:0xee741
        wasm::whitelisted_immutable_slices::test_webgl_rendering_context_immutable_slices::h037a8012a4adaedd@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[514]:0xd190
        core::ops::function::FnOnce::call_once::h19854b0c2793457f@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[1632]:0x784b7
        wasm_bindgen_test::__rt::Context::execute_sync::{{closure}}::h8724c3e402003146@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[1282]:0x5ea03
        <futures::future::lazy::Lazy<F, R>>::get::hb44fd2defcbd4dc2@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[642]:0x20271
        <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll::h06f6de98ca5b4b0b@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[714]:0x244a7
        <wasm_bindgen_test::__rt::TestFuture<F> as futures::future::Future>::poll::{{closure}}::{{closure}}::h5b261ac79651c221@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[1441]:0x6a052
        wasm_bindgen::convert::closures::invoke0_mut::h73ff881cc762e8a8@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[3744]:0xfce1c
        cbarg0@http://127.0.0.1:56658/wasm-bindgen-test:5352:20
        window.__wbg_test_invoke@http://127.0.0.1:56658/:37:38
        __wbg_wbgtestinvoke_f27479ae3a6c5852@http://127.0.0.1:56658/wasm-bindgen-test:5365:13
        wasm_bindgen_test::__rt::__wbg_test_invoke::hf2b2bb34f5c223da@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[3558]:0xf755e
        <wasm_bindgen_test::__rt::TestFuture<F> as futures::future::Future>::poll::{{closure}}::h029fa6859f98969a@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[1369]:0x6722e
        <scoped_tls::ScopedKey<T>>::set::hed98d180e2c04991@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[2189]:0x83ea1
        <wasm_bindgen_test::__rt::TestFuture<F> as futures::future::Future>::poll::he29c70ce5da19936@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[1361]:0x663ef
        <alloc::boxed::Box<F> as futures::future::Future>::poll::h0d1c7651bb0bbb41@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[3729]:0xfbde3
        <wasm_bindgen_test::__rt::ExecuteTests as futures::future::Future>::poll::h4e4a1f89214b76ff@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[3528]:0xf34e7
        <futures::future::map::Map<A, F> as futures::future::Future>::poll::h7062da43a85f3576@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[3960]:0x107769
        <futures::future::map_err::MapErr<A, F> as futures::future::Future>::poll::h6ed2e2a4f31a7a95@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4108]:0x10c5bf
        <alloc::boxed::Box<F> as futures::future::Future>::poll::h9efdadf3e228c51d@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4680]:0x123a29
        <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}::h533ec2b787c19ef8@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4744]:0x12607e
        <futures::task_impl::Spawn<T>>::enter::{{closure}}::hbfde059297b6538d@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4745]:0x1260d0
        futures::task_impl::std::set::hcf3974b140a3b7f1@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4632]:0x12239c
        <futures::task_impl::Spawn<T>>::enter::hd4defc247bcc136c@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4741]:0x125f75
        <futures::task_impl::Spawn<T>>::poll_fn_notify::h29475e2035790e0f@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4740]:0x125e1f
        <futures::task_impl::Spawn<T>>::poll_future_notify::hb03de7d4c5de7247@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4743]:0x126041
        wasm_bindgen_futures::_future_to_promise::Package::poll::h62f8417816046456@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4413]:0x11a3ca
        <wasm_bindgen_futures::_future_to_promise::Package as futures::task_impl::Notify>::notify::{{closure}}::hbdd52fbb3b80720b@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4607]:0x1215e4
        <(dyn core::ops::function::FnMut(A) .> R + 'static) as wasm_bindgen::closure::WasmClosure>::describe::invoke::h066aeafa2f7ee671@http://127.0.0.1:56658/wasm-bindgen-test_bg line 8 > WebAssembly.instantiate:wasm-function[4417]:0x11ad5a
        cb@http://127.0.0.1:56658/wasm-bindgen-test:5736:20

failures:

    wasm::whitelisted_immutable_slices::test_webgl_rendering_context_immutable_slices

test result: FAILED. 54 passed; 1 failed; 0 ignored
console.log div contained:
    Does get context work? null

console.log div contained:
    wasm-bindgen: imported JS function that was not marked as `catch` threw an error:
    TypeError: getObject(...) is null
```

Any ideas?

Reply:
Ah unfortunately I'm not webgl expert so I'm not sure what would cause this :(

Perhaps MDN and/or other resources document reasons why it could return `null` though?

Reply:
Gotcha gotcha - returning `null` usually means that there's no WebGL support for some reason. 

---

> If the browser supports WebGL and canvas.getContext("webgl") returns null then WebGL failed for some reason other than user's browser (no GPU, out of memory, etc...)

[source](https://www.khronos.org/webgl/wiki/FAQ)

---

Alright so I tried using `chromedriver`

`CHROMEDRIVER=chromedriver cargo test --manifest-path crates/web-sys/Cargo.toml --target wasm32-unknown-unknown --all-features`

and tests pass.

However I get back null for the webgl context when using geckodriver

`GECKODRIVER=geckodriver cargo test --manifest-path crates/web-sys/Cargo.toml --target wasm32-unknown-unknown --all-features`

Must be a geckodriver thing.

Reply:
Could one potential approach be to only run the tests for the whitelisted WebGL APIs if the `CHROMEDRIVER` environment variable is set?

So I'd just wrap the WebGL tests in an if statement?

Or is that the wrong approach for some reason(s)?

Reply:
Hm interesting! I wonder if it's something like a capability we need to enable in geckodriver? Some quick googling doesn't turn up much. In any case testing it only in Chrome for now should be fine.